Source/FieldSolver/SpectralSolver/SpectralAlgorithms/PsatdAlgorithmMultiJ.cpp

```diff
@@ -0,0 +1,453 @@
+    C_coef = SpectralRealCoefficients(ba, dm, 1, 0);
+    S_ck_coef = SpectralRealCoefficients(ba, dm, 1, 0);
+    X1_coef = SpectralRealCoefficients(ba, dm, 1, 0);
+    X2_coef = SpectralRealCoefficients(ba, dm, 1, 0);
+    X3_coef = SpectralRealCoefficients(ba, dm, 1, 0);
+
+    InitializeSpectralCoefficients(spectral_kspace, dm, dt);
+
+    // Allocate these coefficients only with time averaging
+    if (time_averaging)
+    {
+        X5_coef = SpectralRealCoefficients(ba, dm, 1, 0);
+        X6_coef = SpectralRealCoefficients(ba, dm, 1, 0);
+        InitializeSpectralCoefficientsAveraging(spectral_kspace, dm, dt);
+    }
+}
+
+void
+PsatdAlgorithmMultiJ::pushSpectralFields (SpectralFieldData& f) const
+{
+    const bool time_averaging = m_time_averaging;
+    const bool dive_cleaning = m_dive_cleaning;
+    const bool divb_cleaning = m_divb_cleaning;
+
+    const amrex::Real dt = m_dt;
+
+    const SpectralFieldIndex& Idx = m_spectral_index;
+
+    // Loop over boxes
+    for (amrex::MFIter mfi(f.fields); mfi.isValid(); ++mfi)
+    {
+        const amrex::Box& bx = f.fields[mfi].box();
+
+        // Extract arrays for the fields to be updated
+        amrex::Array4<Complex> fields = f.fields[mfi].array();
+
+        // These coefficients are always allocated
+        amrex::Array4<const amrex::Real> C_arr = C_coef[mfi].array();
+        amrex::Array4<const amrex::Real> S_ck_arr = S_ck_coef[mfi].array();
+        amrex::Array4<const amrex::Real> X1_arr = X1_coef[mfi].array();
+        amrex::Array4<const amrex::Real> X2_arr = X2_coef[mfi].array();
+        amrex::Array4<const amrex::Real> X3_arr = X3_coef[mfi].array();
+
+        amrex::Array4<const amrex::Real> X5_arr;
+        amrex::Array4<const amrex::Real> X6_arr;
+        if (time_averaging)
+        {
+            X5_arr = X5_coef[mfi].array();
+            X6_arr = X6_coef[mfi].array();
+        }
+
+        // Extract pointers for the k vectors
+        const amrex::Real* modified_kx_arr = modified_kx_vec[mfi].dataPtr();
+#if defined(WARPX_DIM_3D)
+        const amrex::Real* modified_ky_arr = modified_ky_vec[mfi].dataPtr();
+#endif
+        const amrex::Real* modified_kz_arr = modified_kz_vec[mfi].dataPtr();
+
+        // Loop over indices within one box
+        amrex::ParallelFor(bx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept
+        {
+            // Record old values of the fields to be updated
+            const Complex Ex_old = fields(i,j,k,Idx.Ex);
+            const Complex Ey_old = fields(i,j,k,Idx.Ey);
+            const Complex Ez_old = fields(i,j,k,Idx.Ez);
+            const Complex Bx_old = fields(i,j,k,Idx.Bx);
+            const Complex By_old = fields(i,j,k,Idx.By);
+            const Complex Bz_old = fields(i,j,k,Idx.Bz);
+
+            // Shortcuts for the values of J and rho
+            const Complex Jx = fields(i,j,k,Idx.Jx);
+            const Complex Jy = fields(i,j,k,Idx.Jy);
+            const Complex Jz = fields(i,j,k,Idx.Jz);
```

Comment:
We might as well call these `Jx_old`, `Jy_old`, etc. (similarly to the `rho_old`, etc.)

Reply:
Of course this change would need to be propagated throughout the code.